modules/ximgproc/include/opencv2/ximgproc/edge_drawing.hpp

```diff
@@ -0,0 +1,404 @@
+	int noPixels;            // # of pixels making up this circle
+
+							 // If this circle is better approximated by an ellipse, we set isEllipse to true & eq contains the ellipse's equation
+	EllipseEquation eq;
+	double ellipseFitError;  // ellipse fit error
+	bool isEllipse;
+	double majorAxisLength;     // Length of the major axis
+	double minorAxisLength;     // Length of the minor axis
+};
+
+// ------------------------------------------- ARCS ----------------------------------------------------
+struct MyArc {
+	double xc, yc, r;           // center x, y and radius
+	double circleFitError;      // Error during circle fit
+
+	double sTheta, eTheta;      // Start & end angle in radius
+	double coverRatio;          // Ratio of the pixels covered on the covering circle [0-1] (noPixels/circumference)
+
+	int turn;                   // Turn direction: 1 or -1
+
+	int segmentNo;              // SegmentNo where this arc belongs
+
+	int sx, sy;                 // Start (x, y) coordinate
+	int ex, ey;                 // End (x, y) coordinate of the arc
+
+	double *x, *y;              // Pointer to buffer containing the pixels making up this arc
+	int noPixels;               // # of pixels making up the arc
+
+	bool isEllipse;             // Did we fit an ellipse to this arc?
+	EllipseEquation eq;         // If an ellipse, then the ellipse's equation
+	double ellipseFitError;     // Error during ellipse fit
+};
+
+// =============================== AngleSet ==================================
+
+// add a circular arc to the list of arcs
+inline double ArcLength(double sTheta, double eTheta)
+{
+	if (eTheta > sTheta)
+        return eTheta - sTheta;
+	else
+        return CV_2PI - sTheta + eTheta;
+}
+
+// A fast implementation of the AngleSet class. The slow implementation is really bad. About 10 times slower than this!
+struct AngleSetArc {
+	double sTheta;
+	double eTheta;
+	int next;         // Next AngleSetArc in the linked list
+};
+
+struct AngleSet {
+	AngleSetArc angles[360];
+	int head;
+	int next;   // Next AngleSetArc to be allocated
+	double overlapAmount;   // Total overlap of the arcs in angleSet. Computed during set() function
+
+	AngleSet() { clear(); } //end-AngleSet
+	void clear() { head = -1; next = 0; overlapAmount = 0; }
+	double overlapRatio() { return overlapAmount / CV_2PI; }
+
+	void _set(double sTheta, double eTheta);
+	void set(double sTheta, double eTheta);
+
+	double _overlap(double sTheta, double eTheta);
+	double overlap(double sTheta, double eTheta);
+
+	void computeStartEndTheta(double &sTheta, double &eTheta);
+	double coverRatio();
+};
+
+
+struct EDArcs {
+	MyArc *arcs;
+	int noArcs;
+
+public:
+	EDArcs(int size = 10000) {
+		arcs = new MyArc[size];
+		noArcs = 0;
+	} //end-EDArcs
+
+	~EDArcs() {
+		delete arcs;
+	} //end-~EDArcs
+};
+
+
+//-----------------------------------------------------------------
+// Buffer manager
+struct BufferManager {
+	double *x, *y;
+	int index;
```

Comment:
Looks like there are too many implementation details in Public API.
Lets keep Public API minimal (keep only necessary parts for samples / tests)

Move other code to `src/*.hpp` / `src/*.cpp` files.